### PR TITLE
#911 enable findbugs

### PR DESCRIPTION
**Pull request by exper0**
#911 enable findbugs

### CODE REVIEW

**Comment by dmzaytsev** [comment]
@exper0 try to fix some findbugs issues until your time budget is over
then create a puzzle for others
Removing exludes from POM-file not enough to complete the ticket

**Comment by karato** [comment]
@exper0 Many thanks, I will find someone to review it before we merge

**Comment by karato** [comment]
@dmzaytsev review this pull request please, it is rather short for a full review

**Comment by exper0** [comment]
@dmzaytsev how should I create puzzle for other developers? I fixed pom.xml, there are too many places in the code which require fixing. Where should I created `@todo`?

File: netbout-client/pom.xml [hunk 1]
```diff
@@ -178,11 +178,6 @@
                     <plugin>
                         <groupId>com.qulice</groupId>
                         <artifactId>qulice-maven-plugin</artifactId>
-                        <configuration>
-                            <excludes>
-                                <exclude>findbugs:.*</exclude>
```
Comment (dmzaytsev) [inline netbout-client/pom.xml]: @exper0  right now we have whole project excluded from the FindBugs checking `<exclude>findbugs:.*</exclude>`
you have to fix some files and and have to set exclusions for others.
then create a new puzzle here to fix other files and remove them from the exclusions

**Comment by dmzaytsev** [comment]
@exper0 see my answer above
we need green build before merge

File: netbout-client/pom.xml (hunk 1 above)
Comment (dmzaytsev) [inline netbout-client/pom.xml]: @exper0 you can use regexp for the exclusions

**Comment by exper0** [comment]
@dmzaytsev puzzle is just `@todo` comment in the code, right? where exactly should create it?

File: netbout-client/pom.xml (hunk 1 above)
Comment (exper0) [inline netbout-client/pom.xml]: @dmzaytsev thanks, it's clear. But what about puzzle creation? see below question

**Comment by dmzaytsev** [comment]
@exper0 right, you can create it after `<excludes>` tag for instance (line 183)

**Comment by dmzaytsev** [comment]
@exper0 this is example of  a puzzle in pom.xml https://github.com/libreio/libre/blob/master/pom.xml#L413

**Comment by exper0** [comment]
@dmzaytsev many thanks, now it's clear. For some reason I thought that @todo should be in the code :)

**Comment by exper0** [comment]
@dmzaytsev looking into findbugs erros. Actually, almost all of them doesn't worth fixing. Wast majority of them is just about creating private static classes instead of inline classes (`new SomeInterface ...`). I'd say it makes sense to tune findbugs filters at first. Also, findbugs itself is pretty old and has its own bugs - worth updaing. So virtually all bugs remain unfixed - should I just name each class in exclude or just return `*`?

**Comment by exper0** [comment]
@dmzaytsev one more observation - qulice exclude filter for findbug just doesn't work - it always send empty string as class name so it's impossible to make any other that 'exclude all'

**Comment by exper0** [comment]
@dmzaytsev finally done as you suggested - fixed some issues and created puzzles for the others

File: netbout-client/pom.xml [hunk 2]
```diff
@@ -180,7 +180,12 @@
                         <artifactId>qulice-maven-plugin</artifactId>
                         <configuration>
                             <excludes>
-                                <exclude>findbugs:.*</exclude>
+                                <!--
+                                @todo #911 Fix findbugs errors in client
```
Comment (dmzaytsev) [inline netbout-client/pom.xml]: @exper0  the puzzle description should contains at least 20 words, also you have to use format like this
`@todo #911 :30min Fix findbugs errors in client`
I wonder why Travis's build succeeded

File: netbout-spi/pom.xml [hunk 3]
```diff
@@ -90,7 +90,13 @@
                         <artifactId>qulice-maven-plugin</artifactId>
                         <configuration>
                             <excludes>
-                                <exclude>findbugs:.*</exclude>
+                                <!--
+                                @todo #911 Fix findbugs errors in SPI classes
```
Comment (dmzaytsev) [inline netbout-spi/pom.xml]: @exper0 please see the comment above about the  format of the puzzle

File: netbout-web/pom.xml [hunk 4]
```diff
@@ -736,6 +736,11 @@
                                 <exclude>checkstyle:/src/main/scss/.*</exclude>
                                 <exclude>checkstyle:/src/test/casperjs/.*</exclude>
                                 <exclude>checkstyle:/src/main/js/.*</exclude>
+                                <!--
+                                @todo #911 Fix findbugs errors in web
```
Comment (dmzaytsev) [inline netbout-web/pom.xml]: @exper0 please see the comment above about the  format of the puzzle

File: netbout-web/pom.xml [hunk 5]
```diff
@@ -736,6 +736,11 @@
...
+                                 filters in qulice 0.12 are broken, right excludes
+                                 should be more granular
+                                -->
                                 <exclude>findbugs:.*</exclude>
```
Comment (dmzaytsev) [inline netbout-web/pom.xml]: @exper0 btw I would suggest you create a ticket in Qulice project
perhaps it is a bug 
If they accept the bug you can put link to it here

File: netbout-web/src/main/java/com/netbout/rest/MarkdownPegdown.java [hunk 6]
```diff
@@ -156,7 +158,12 @@ private static Tidy makeTidy() {
         final Tidy tidy = new Tidy();
         tidy.setShowErrors(0);
         tidy.setErrout(
-            new PrintWriter(Logger.stream(Level.FINE, MarkdownPegdown.class))
+            new PrintWriter(
```
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/rest/MarkdownPegdown.java]: @exper0 btw what exactly Findbugs says about it?
PrintWriter has OutputStream in its constructor and Logger.stream returns OutputStream too

**Comment by dmzaytsev** [comment]
@exper0 see my comments above
but  I suggest wait unitil we update Qulice dependence

**Comment by dmzaytsev** [comment]
@karato we are waiting for  #909

File: netbout-client/pom.xml (hunk 2 above)
Comment (exper0) [inline netbout-client/pom.xml]: @dmzaytsev whole puzzle description is bigger than a single line. I have no idea what to write here. Is it somehow unclear? Formatting is clear, will fix it

File: netbout-web/pom.xml (hunk 5 above)
Comment (exper0) [inline netbout-web/pom.xml]: @dmzaytsev I created a ticket but suggested to try newer qulice version. In newer version(0.15.2) it is fixed. The problem is that Netbout uses 0.12

File: netbout-web/src/main/java/com/netbout/rest/MarkdownPegdown.java (hunk 6 above)
Comment (exper0) [inline netbout-web/src/main/java/com/netbout/rest/MarkdownPegdown.java]: @dmzaytsev it was about relying on default charset

File: netbout-client/pom.xml (hunk 2 above)
Comment (dmzaytsev) [inline netbout-client/pom.xml]: @exper0 thr puzzle can be placed on several lines

Comment (exper0) [inline netbout-client/pom.xml]: @dmzaytsev this is what I'm saying - this puzzle is bigger than 1 line. I just trying to understand what's wrong except absence of ":30min"

Comment (dmzaytsev) [inline netbout-client/pom.xml]: @exper0 it contains 17 words
Rultor will reject puzzles that less than 20 words

**Comment by exper0** [comment]
@dmzaytsev fixed puzzles

**Comment by dmzaytsev** [comment]
@exper0 thanks
we will wait until #909 resolved

**Comment by karato** [comment]
> @karato we are waiting for  #909 

@dmzaytsev yes, waiting for #909

**Comment by karato** [comment]
@exper0 impediment in issue #909 is closed, we can probably try to continue here

**Comment by exper0** [comment]
@dmzaytsev @karato not sure how this one and #909 are connected. From my perspective I've finished the work on #911 in this PR.

**Comment by dmzaytsev** [comment]
@exper0 let's wait until a new ticket puzzle will be created

**Comment by exper0** [comment]
@karato impediment is outdated qulice - it was not upgrade in scope of #909. Please create separate ticket to cover puzzle created in scope of it.

**Comment by karato** [comment]
> @karato impediment is outdated qulice - it was not upgrade in scope of #909. Please create separate ticket to cover puzzle created in scope of it.

@exper0 I'll do it automatically, when the next deployment cycle passes

File: netbout-client/pom.xml [hunk 7]
```diff
@@ -180,7 +180,13 @@
                         <artifactId>qulice-maven-plugin</artifactId>
                         <configuration>
                             <excludes>
-                                <exclude>findbugs:.*</exclude>
+                                <!--
+                                @todo #911:30min Fix findbugs errors in client
+                                 Some issues may not require fixing an should be
+                                 excluded. Filters in qulice 0.12 are broken, right
```
Comment (dmzaytsev) [inline netbout-client/pom.xml]: @exper0 did you create an issue on Qulice project for that?
put link to the issue if yes

Comment (exper0) [inline netbout-client/pom.xml]: @dmzaytsev I created and issue but after discussion with the team I closed it - due to the fact that issue I found is not reproducable in version 0.15.2 (Netbout uses 0.12)

**Comment by exper0** [comment]
@dmzaytsev issue with findbugs filters in qulice 0.12 is pretty obvious - I can observe *xml files with findbugs configuration created by Maven plugin - it contains something like

```
       <Class name="null" />
```

**Comment by exper0** [comment]
@dmzaytsev I'd propose to update qulice explicitly not waiting for jcabi-parent upgrade (jcabi-parent also contains outdated version of qulice). I created [issue about](#961). What do yo think?

File: netbout-client/pom.xml (hunk 7 above)
Comment (dmzaytsev) [inline netbout-client/pom.xml]: @exper0 let's merge this PR, please update the puzzle
I think it should be resolved when #961 completed due to broken Findbugs filters in current  version of Qulice

**Comment by exper0** [comment]
@dmzaytsev done.

**Comment by dmzaytsev** [comment]
@exper0 thanks

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by karato** [comment]
@dmzaytsev **36 mins** sent to your balance (ID `75307242`), many thanks! It took 146 hours and 44 mins.; extra minutes for review comments (c=21); +36 to your rating, your total score is [+2522](http://www.netbout.com/b/35166?open=rating)

**Comment by karato** [comment]
@rultor deploy

**Pull request merged**